README.md

![Build Status](https://github.com/Quid2/flat/actions/workflows/haskell-ci.yml/badge.svg)

[![Hackage version](https://img.shields.io/hackage/v/flat.svg)](http://hackage.haskell.org/package/flat)

[![Stackage LTS 16](http://stackage.org/package/flat/badge/lts-16)](http://stackage.org/lts/package/flat)
[![Stackage LTS 18](http://stackage.org/package/flat/badge/lts-18)](http://stackage.org/lts/package/flat)
[![Stackage LTS 19](http://stackage.org/package/flat/badge/lts-19)](http://stackage.org/lts/package/flat)
[![Stackage Nightly](http://stackage.org/package/flat/badge/nightly)](http://stackage.org/nightly/package/flat)

Haskell implementation of [Flat](http://quid2.org/docs/Flat.pdf), a principled, portable and compact binary data format ([specs](http://quid2.org)).


### How To Use It For Fun and Profit

```haskell
{-# LANGUAGE DeriveGeneric, DeriveAnyClass #-}
```

```haskell
import Flat
```



```haskell
data Direction = North | South | Center | East | West deriving (Show,Generic,Flat)
```

Use **flat** to encode: 

```haskell
flat [North,South]
-> "\149"
```


and **unflat** to decode:

```haskell
unflat (flat [North,South]) :: Decoded [Direction]
-> Right [ North , South ]
```


And thanks to Flat's bit-encoding, this list fits in 1 byte (rather than the 5 bytes that would be required by a traditional byte encoding):

```haskell
flatBits [North,South]
-> "10010101"
```



### Performance

For some hard data, see this [comparison of the major haskell serialisation libraries](https://github.com/haskell-perf/serialization).

Briefly:
 * Size: `flat` produces significantly smaller binaries than all other libraries (3/4 times usually)
 * Serialization time: `store`, `persist` and `flat` are faster
 * Deserialization time: `store`, `flat`, `persist` and `cereal` are faster
 * Transfer time (serialisation time + transport time on the network + deserialisation at the receiving end): `flat` is usually faster for all but the highest network speeds

### Documentation

* [Tutorial](http://hackage.haskell.org/package/flat/docs/Flat-Tutorial.html)

* [Hackage Package and Docs](http://hackage.haskell.org/package/flat)

* [Flat Format Specification](http://quid2.org/docs/Flat.pdf)

### Installation

Get the latest stable version from [hackage](https://hackage.haskell.org/package/flat).

### Compatibility

Tested with:

* [GHC](https://www.haskell.org/ghc/) 7.10.3 to 9.4.3 (x64)


* [GHCJS](https://github.com/ghcjs/ghcjs) version 8.6.0.1 (GHC 8.6.2) 



### Known Bugs and Infelicities

* Data types with more than 512 constructors are currently unsupported (but support could be easily added if necessary)

* Longish compilation times

  * To improve performance, `flat` relies on extensive inlining. This unfortunately leads to longer compilation times.

    If you have many data types or very large ones, you might want to: 

      * During development, turn optimisations off (`stack --fast` or `-O0` in the cabal file).

      * Keep your serialisation code in separate modules.

* See also the [full list of open issues](https://github.com/Quid2/flat/issues).

### Ports for other languages

[Rust](https://www.rust-lang.org/) and [TypeScript-JavaScript](https://github.com/Quid2/ts) ports are under development.


Get in touch if you would like to help porting `flat` to other languages.

### Acknowledgements

`flat` reuses ideas and readapts code from various packages, mainly: `store`, `binary-bits` and `binary` and includes bug fixes from a number of contributors.

### Other Stuff You Might Like

To decode `flat` encoded data you need to know the type of the serialised data.

This is ok for applications that do not require long-term storage and that do not operate in open distributed systems.

For those who do, you might want to supplement `flat` with [ZM - Language independent, reproducible, absolute types](https://github.com/Quid2/zm).


CHANGELOG
Significant and compatibility-breaking changes.


Version 0.6:
    - [**breaking**] Renamed Repr to AsBin and SizeOf to AsSize
    - Faster AsBin implementation 
    - Fixed AsSize implementation (SizeOf worked only for types with custom size implementation and returned max size, not actual size)

Version 0.5.2:
    - Fixed https://github.com/Quid2/flat/issues/28 that could cause an undetected overflow when decoding out of range Int/Word values
    - Added UTF16Text and Data.ByteString.Short support for ghcjs
    - Improved speed and accuracy of encoded Text size calculation (for encoding)
    - [**breaking**] Removed `textBytes` from Flat.Encoder.Size (breaking but insignificant)

Version 0.5:
    - Compatibility with ghc 9.0.2 & 9.2.4 & 9.4.2
    - Compatibility with text-2.0 (GHC)
    - Fixed https://github.com/Quid2/flat/issues/23 that could cause an encoding failure for non byte aligned Arrays 
    - Fixed https://github.com/Quid2/flat/pull/26 that could cause the decoder to read beyond the end of the decoding buffer (causing a SEGFAULT on ghcjs)
    - Merged https://github.com/Quid2/flat/pull/22 - fails correctly on invalid UTF-8 text 
    - Added a few ways to partially or incrementally decode values, see Flat.Repr and Flat.Decoder.Run.listTDecoder
    - Moved strictDecoder from Flat.Decoder.Types to Flat.Decoder
    - Removed Flat instance for Semigroup.Option from Flat.Instances.Base
    - Moved overlapping/specialised Flat instance for [Char] to Flat.Instances.Extra
    - Added a few extra functions

Version 0.4.4:
	- Added instances for Identity, Monoid.Dual/All/Any/Sum/Product/Alt, Semigroup.Min/Max/First/Last/Option 

Version 0.4.2:
	- Fixed links in README.md
	- Added multiple stack.yaml files for different LTS

Version 0.4:
    - Compatibility with ghc 8.8.3
	- Data.Seq.Seq is serialised as a List (previously was serialised as an Array)
	- Changed namespace from Data.Flat to Flat
	- Addtional Flat Instances for some common packages: array, base, bytestring, containers, dlist, text, unordered-containers, vector
	- Additional doctests, with a static version that can run without ghci and therefore also under ghcjs/eta (run it with stack test :doc-static)
	- Many minor changes

Version 0.3.4:
	- Redisegned Generics-based instance generation (Data.Flat.Class) to reduce compilation time and improve encoding and decoding performance
	- Fixed GHCJS Double bug and tested GHCJS with full test suite

Version 0.3.2:
	- Tested with ghc 8.2.1
	- Dropped dependencies on the 'cpu', 'derive' and 'tasty' packages to make it compatible with the Eta compiler (https://github.com/typelead/eta)

Version 0.3:
	- Removed 'flatStrict' and 'unflatStrict' (use 'flat' and 'unflat' instead that also encode/decode strictly)
	- `unflatWith` now takes a decoder for the unpadded value (previously it expected a padded decoder) and decodes the padded value
	- Added some decoding primitives
	- Added Data.ByteString.Convert


src/README-EXE.hs
{-# LANGUAGE FlexibleInstances ,UndecidableInstances , NoMonomorphismRestriction #-}
-- ![Build Status](https://github.com/Quid2/flat/actions/workflows/haskell-ci.yml/badge.svg)

-- [![Hackage version](https://img.shields.io/hackage/v/flat.svg)](http://hackage.haskell.org/package/flat)

-- <!--
-- [![Stackage LTS 6](http://stackage.org/package/flat/badge/lts-6)](http://stackage.org/lts/package/flat)
-- [![Stackage LTS 9](http://stackage.org/package/flat/badge/lts-9)](http://stackage.org/lts/package/flat)
-- [![Stackage LTS 11](http://stackage.org/package/flat/badge/lts-11)](http://stackage.org/lts/package/flat)
-- [![Stackage LTS 12](http://stackage.org/package/flat/badge/lts-12)](http://stackage.org/lts/package/flat)
-- [![Stackage LTS 14](http://stackage.org/package/flat/badge/lts-14)](http://stackage.org/lts/package/flat)
-- -->
-- [![Stackage LTS 16](http://stackage.org/package/flat/badge/lts-16)](http://stackage.org/lts/package/flat)
-- [![Stackage LTS 18](http://stackage.org/package/flat/badge/lts-18)](http://stackage.org/lts/package/flat)
-- [![Stackage LTS 19](http://stackage.org/package/flat/badge/lts-19)](http://stackage.org/lts/package/flat)
-- [![Stackage Nightly](http://stackage.org/package/flat/badge/nightly)](http://stackage.org/nightly/package/flat)

-- <!--
-- -->
-- Haskell implementation of [Flat](http://quid2.org/docs/Flat.pdf), a principled, portable and compact binary data format ([specs](http://quid2.org)).


-- ### How To Use It For Fun and Profit

{-# LANGUAGE DeriveGeneric, DeriveAnyClass #-}
{-# LANGUAGE IncoherentInstances, NoMonomorphismRestriction #-} -- DELETE
import qualified Data.Text as T
import qualified Text.Show.Pretty
import Text.PrettyPrint.HughesPJClass(prettyShow) -- DELETE
import Flat.Bits(bits,paddedBits) -- DELETE

import Flat

p :: Flat a => a -> String -- DELETE
p = prettyShow . bits -- DELETE

flatBits = prettyShow . paddedBits -- DELETE

data Direction = North | South | Center | East | West deriving (Show,Generic,Flat)

-- Use **flat** to encode: 

d3 =  flat [North,South]

-- and **unflat** to decode:

d4 =  unflat (flat [North,South]) :: Decoded [Direction]

-- And thanks to Flat's bit-encoding, this list fits in 1 byte (rather than the 5 bytes that would be required by a traditional byte encoding):

d5 =  flatBits [North,South]


-- ### Performance

-- For some hard data, see this [comparison of the major haskell serialisation libraries](https://github.com/haskell-perf/serialization).

-- Briefly:
--  * Size: `flat` produces significantly smaller binaries than all other libraries (3/4 times usually)
--  * Serialization time: `store`, `persist` and `flat` are faster
--  * Deserialization time: `store`, `flat`, `persist` and `cereal` are faster
--  * Transfer time (serialisation time + transport time on the network + deserialisation at the receiving end): `flat` is usually faster for all but the highest network speeds

-- ### Documentation

-- * [Tutorial](http://hackage.haskell.org/package/flat/docs/Flat-Tutorial.html)

-- * [Hackage Package and Docs](http://hackage.haskell.org/package/flat)

-- * [Flat Format Specification](http://quid2.org/docs/Flat.pdf)

-- ### Installation

-- Get the latest stable version from [hackage](https://hackage.haskell.org/package/flat).

-- ### Compatibility

-- Tested with:

-- * [GHC](https://www.haskell.org/ghc/) 7.10.3 to 9.4.2 (x64)

-- <!--
-- * [GHC](https://www.haskell.org/ghc/) 7.10.3/LLVM 3.5.2 (Arm7)
--   * Caveat: not recently tested
-- -->

-- * [GHCJS](https://github.com/ghcjs/ghcjs) version 8.6.0.1 (GHC 8.6.2) 

--   * Notes:
--     * versions prior to 0.3.4 encode `Double` values incorrectly when they are not aligned with a byte boundary.
--     * versions prior to 0.6 did not support short bytestrings (Data.ByteString.Short) and UTF16Text

-- <!--

-- * [ETA](https://eta-lang.org/) 0.8.6b2 & etlas>=1.5.0.0
--   * Caveats:
--     * Compatible since version 0.36
--     * Eta requires patched versions of certain packages. When compiling under eta, `flat`'s cabal file selects the latest patched versions available at the time of release.
--     * Eta has an issue with recursion, that in certain cases leads to a st There is a problem with recursion. This causes teh encoding of lists larger than about 1500 elements (or whatever the java maximum stack is set to be), to fail.
--   See [issue](https://github.com/typelead/eta/issues/901) for details.
-- -->


-- ### Known Bugs and Infelicities

-- * Data types with more than 512 constructors are currently unsupported (but support could be easily added if necessary)

-- * Longish compilation times

--   * `flat` relies more than other serialisation libraries on extensive inlining for its good performance, this unfortunately leads to longer compilation times. 

--     If you have many data types or very large ones this might become an issue.

--     A couple of good practices that will eliminate or mitigate this problem are:

--       * During development, turn optimisations off (`stack --fast` or `-O0` in the cabal file).

--       * Keep your serialisation code in a separate module or modules.

-- * See also the [full list of open issues](https://github.com/Quid2/flat/issues).

-- ### Ports for other languages

-- [Rust](https://www.rust-lang.org/) and [TypeScript-JavaScript](https://github.com/Quid2/ts) ports are under development.

-- <!--
-- and [Purescript](https://www.purescript.org/) 
-- -->

-- Get in touch if you would like to help porting `flat` to other languages.

-- ### Acknowledgements

-- `flat` reuses ideas and readapts code from various packages, mainly: `store`, `binary-bits` and `binary` and includes bug fixes from a number of contributors.

-- ### Other Stuff You Might Like

-- To decode `flat` encoded data you need to know the type of the serialised data.

-- This is ok for applications that do not require long-term storage and that do not operate in open distributed systems.

-- For those who do, you might want to supplement `flat` with [ZM - Language independent, reproducible, absolute types](https://github.com/Quid2/zm).
class IO_ a where io_  :: a -> IO String
instance {-# OVERLAPPABLE #-} Show a => IO_ a where io_  = return . showRes
instance {-# OVERLAPPING #-} IO_ T.Text where io_  = return . showText
instance {-# OVERLAPPING  #-} Show a => IO_ (IO a) where io_ = fmap showRes
showRes = unlines . map ("-> " ++) . lines . Text.Show.Pretty.ppShow
showText = unlines . map ("-- " ++) . lines . T.unpack
main = sequence [return "\n-- ![Build Status](https://github.com/Quid2/flat/actions/workflows/haskell-ci.yml/badge.svg)",return "\n",return "\n-- [![Hackage version](https://img.shields.io/hackage/v/flat.svg)](http://hackage.haskell.org/package/flat)",return "\n",return "\n-- <!--",return "\n-- [![Stackage LTS 6](http://stackage.org/package/flat/badge/lts-6)](http://stackage.org/lts/package/flat)",return "\n-- [![Stackage LTS 9](http://stackage.org/package/flat/badge/lts-9)](http://stackage.org/lts/package/flat)",return "\n-- [![Stackage LTS 11](http://stackage.org/package/flat/badge/lts-11)](http://stackage.org/lts/package/flat)",return "\n-- [![Stackage LTS 12](http://stackage.org/package/flat/badge/lts-12)](http://stackage.org/lts/package/flat)",return "\n-- [![Stackage LTS 14](http://stackage.org/package/flat/badge/lts-14)](http://stackage.org/lts/package/flat)",return "\n-- -->",return "\n-- [![Stackage LTS 16](http://stackage.org/package/flat/badge/lts-16)](http://stackage.org/lts/package/flat)",return "\n-- [![Stackage LTS 18](http://stackage.org/package/flat/badge/lts-18)](http://stackage.org/lts/package/flat)",return "\n-- [![Stackage LTS 19](http://stackage.org/package/flat/badge/lts-19)](http://stackage.org/lts/package/flat)",return "\n-- [![Stackage Nightly](http://stackage.org/package/flat/badge/nightly)](http://stackage.org/nightly/package/flat)",return "\n",return "\n-- <!--",return "\n-- -->",return "\n-- Haskell implementation of [Flat](http://quid2.org/docs/Flat.pdf), a principled, portable and compact binary data format ([specs](http://quid2.org)).",return "\n",return "\n",return "\n-- ### How To Use It For Fun and Profit",return "\n",return "\n{-# LANGUAGE DeriveGeneric, DeriveAnyClass #-}",return "\n{-# LANGUAGE IncoherentInstances, NoMonomorphismRestriction #-} -- DELETE",return "\nimport Text.PrettyPrint.HughesPJClass(prettyShow) -- DELETE",return "\nimport Flat.Bits(bits,paddedBits) -- DELETE",return "\n",return "\nimport Flat",return "\n",return "\np :: Flat a => a -> String -- DELETE",return "\np = prettyShow . bits -- DELETE",return "\n",return "\nflatBits = prettyShow . paddedBits -- DELETE",return "\n",return "\ndata Direction = North | South | Center | East | West deriving (Show,Generic,Flat)",return "\n",return "\n-- Use **flat** to encode: ",return "\n",return "\nflat [North,South]",return "\n",io_ d3,return "\n",return "\n-- and **unflat** to decode:",return "\n",return "\nunflat (flat [North,South]) :: Decoded [Direction]",return "\n",io_ d4,return "\n",return "\n-- And thanks to Flat's bit-encoding, this list fits in 1 byte (rather than the 5 bytes that would be required by a traditional byte encoding):",return "\n",return "\nflatBits [North,South]",return "\n",io_ d5,return "\n",return "\n",return "\n-- ### Performance",return "\n",return "\n-- For some hard data, see this [comparison of the major haskell serialisation libraries](https://github.com/haskell-perf/serialization).",return "\n",return "\n-- Briefly:",return "\n--  * Size: `flat` produces significantly smaller binaries than all other libraries (3/4 times usually)",return "\n--  * Serialization time: `store`, `persist` and `flat` are faster",return "\n--  * Deserialization time: `store`, `flat`, `persist` and `cereal` are faster",return "\n--  * Transfer time (serialisation time + transport time on the network + deserialisation at the receiving end): `flat` is usually faster for all but the highest network speeds",return "\n",return "\n-- ### Documentation",return "\n",return "\n-- * [Tutorial](http://hackage.haskell.org/package/flat/docs/Flat-Tutorial.html)",return "\n",return "\n-- * [Hackage Package and Docs](http://hackage.haskell.org/package/flat)",return "\n",return "\n-- * [Flat Format Specification](http://quid2.org/docs/Flat.pdf)",return "\n",return "\n-- ### Installation",return "\n",return "\n-- Get the latest stable version from [hackage](https://hackage.haskell.org/package/flat).",return "\n",return "\n-- ### Compatibility",return "\n",return "\n-- Tested with:",return "\n",return "\n-- * [GHC](https://www.haskell.org/ghc/) 7.10.3 to 9.4.2 (x64)",return "\n",return "\n-- <!--",return "\n-- * [GHC](https://www.haskell.org/ghc/) 7.10.3/LLVM 3.5.2 (Arm7)",return "\n--   * Caveat: not recently tested",return "\n-- -->",return "\n",return "\n-- * [GHCJS](https://github.com/ghcjs/ghcjs) version 8.6.0.1 (GHC 8.6.2) ",return "\n",return "\n--   * Notes:",return "\n--     * versions prior to 0.3.4 encode `Double` values incorrectly when they are not aligned with a byte boundary.",return "\n--     * versions prior to 0.6 did not support short bytestrings (Data.ByteString.Short) and UTF16Text",return "\n",return "\n-- <!--",return "\n",return "\n-- * [ETA](https://eta-lang.org/) 0.8.6b2 & etlas>=1.5.0.0",return "\n--   * Caveats:",return "\n--     * Compatible since version 0.36",return "\n--     * Eta requires patched versions of certain packages. When compiling under eta, `flat`'s cabal file selects the latest patched versions available at the time of release.",return "\n--     * Eta has an issue with recursion, that in certain cases leads to a st There is a problem with recursion. This causes teh encoding of lists larger than about 1500 elements (or whatever the java maximum stack is set to be), to fail.",return "\n--   See [issue](https://github.com/typelead/eta/issues/901) for details.",return "\n-- -->",return "\n",return "\n",return "\n-- ### Known Bugs and Infelicities",return "\n",return "\n-- * Data types with more than 512 constructors are currently unsupported (but support could be easily added if necessary)",return "\n",return "\n-- * Longish compilation times",return "\n",return "\n--   * `flat` relies more than other serialisation libraries on extensive inlining for its good performance, this unfortunately leads to longer compilation times. ",return "\n",return "\n--     If you have many data types or very large ones this might become an issue.",return "\n",return "\n--     A couple of good practices that will eliminate or mitigate this problem are:",return "\n",return "\n--       * During development, turn optimisations off (`stack --fast` or `-O0` in the cabal file).",return "\n",return "\n--       * Keep your serialisation code in a separate module or modules.",return "\n",return "\n-- * See also the [full list of open issues](https://github.com/Quid2/flat/issues).",return "\n",return "\n-- ### Ports for other languages",return "\n",return "\n-- [Rust](https://www.rust-lang.org/) and [TypeScript-JavaScript](https://github.com/Quid2/ts) ports are under development.",return "\n",return "\n-- <!--",return "\n-- and [Purescript](https://www.purescript.org/) ",return "\n-- -->",return "\n",return "\n-- Get in touch if you would like to help porting `flat` to other languages.",return "\n",return "\n-- ### Acknowledgements",return "\n",return "\n-- `flat` reuses ideas and readapts code from various packages, mainly: `store`, `binary-bits` and `binary` and includes bug fixes from a number of contributors.",return "\n",return "\n-- ### Other Stuff You Might Like",return "\n",return "\n-- To decode `flat` encoded data you need to know the type of the serialised data.",return "\n",return "\n-- This is ok for applications that do not require long-term storage and that do not operate in open distributed systems.",return "\n",return "\n-- For those who do, you might want to supplement `flat` with [ZM - Language independent, reproducible, absolute types](https://github.com/Quid2/zm)."] >>= writeFile "/Users/titto/workspace/flat/src/README-OUT.hs" . concat
